features/telemetry/protocols/collection/src/main/java/org/opennms/netmgt/telemetry/protocols/collection/AbstractPersistingAdapter.java

```diff
@@ -188,10 +203,46 @@ public void handleMessageLog(TelemetryMessageLog messageLog) {
                 LOG.trace("Persisting collection set: {} for message: {}", collectionSet, message);
                 final Persister persister = persisterFactory.createPersister(EMPTY_SERVICE_PARAMETERS, repository);
                 collectionSet.visit(persister);
+
+                // Thresholding
+                try {
+                    if (isThresholdingEnabled) {
+                        ThresholdingSession session = getSessionForAgent(result.getAgent(), repository);
+                        session.accept(collectionSet);
+                    }
+                } catch (ThresholdInitializationException e) {
+                    LOG.warn("Failed Thresholding of CollectionSet: {} ", e.getMessage());
+                }
             });
         }
     }
 
+    private ThresholdingSession getSessionForAgent(CollectionAgent agent, RrdRepository repository) throws ThresholdInitializationException {
+        if (thresholdingService == null) {
+            // If we don't have a ThresholdingService,
+            // we are running in the OSGi container (i.e. on a Sentinal) with no Thresholding Service Configured
+            // Disable Thresholding.
+            isThresholdingEnabled = false;
+            throw new ThresholdInitializationException("No ThresholdingService available. No future Threshholding will be done");
+        }
+        // Map of sessions keyed by agent
+        int nodeId = agent.getNodeId();
+        String hostAddress = agent.getHostAddress();
+        String serviceName = adapterConfig.getName();
+        String sessionKey = getSessionKey(nodeId, hostAddress, serviceName);
+
+        ThresholdingSession session = agentThresholdingSessions.asMap().get(sessionKey);
```

Comment:
Do we need to call `asMap()` for every cache lookup?

Reply:
> asMap().get(key) is essentially equivalent to cache.getIfPresent(key), and never causes values to be loaded. 

I will change it to getIfPresent(key) though for readability